balloon/src/params.rs

```diff
@@ -0,0 +1,108 @@
+//! Argon2 password hash parameters.
+
+use crate::{Error, Result};
+use core::num::NonZeroU32;
+#[cfg(feature = "password-hash")]
+use {
+    core::convert::TryFrom,
+    password_hash::{errors::InvalidValue, ParamsString, PasswordHash},
+};
+
+/// Argon2 password hash parameters.
+///
+/// These are parameters which can be encoded into a PHC hash string.
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub struct Params {
+    /// Space cost, expressed in of blocks.
+    pub s_cost: NonZeroU32,
+    /// Time cost, expressed in number of rounds.
+    pub t_cost: NonZeroU32,
+    /// Degree of parallelism, expressed in number of threads.
+    pub p_cost: NonZeroU32,
+}
+
+impl Params {
+    /// Default memory cost.
+    pub const DEFAULT_S_COST: u32 = 1024;
+
+    /// Default number of iterations (i.e. "time").
+    pub const DEFAULT_T_COST: u32 = 1;
+
+    /// Default degree of parallelism.
+    pub const DEFAULT_P_COST: u32 = 1;
```

Comment:
I just used the defaults from the reference implementation, this is far from optimal though. Any guidance would be appreciated.

Reply:
I think you can get rid of the `new()` function and just leave the struct fields as public. The function doesn't do any bounds checking that isn't already enforced by the type system

Reply:
The `Params` fields are already public, I made the `new` function to make it easier to create a `Params` struct, otherwise it would need `NonZeroU32::new(x).unwrap()` multiple times.

Reply:
So after some reading in the paper, I understood that compared to Argon2i, the space cost should be one quarter and the time cost should be the same. So I adjusted the defaults according to the defaults we set in Argon2.

Any other suggestions are welcome.